examples/wmt/models.py

```diff
@@ -0,0 +1,774 @@
+# Lint as: python3
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer-based machine translation model."""
```

Comment:
This is a full encoder-decoder Transformer - i.e. the original "Transformer".  Instead of just using a stack of decoder-layers as in the LM1B example it first has a stack of encoders from the source-language that creates an encoded representation that is then used to "guide" a _mixed_ stack of decoder and "encoder-decoder" layers acting on the target-language.  The primary difference between these different kinds of layers is the masking used to enforce causality as well as some zero-padding masks.  So although they're raw-ingredients are similar (the attention mechanism), the details are quite different between the two.  
It is possible to do translation w. a pure language-model by treating the input sentence as a prompt, but these translation systems tend to be less parameter-efficient and harder to train to SotA performance.